task/buildah/0.1/buildah.yaml

```diff
@@ -15,350 +16,354 @@ spec:
+      description: The counting of Java components by publisher in JSON format
+      type: string
+    - name: JAVA_COMMUNITY_DEPENDENCIES
+      description: The Java dependencies that came from community sources such as Maven central.
   stepTemplate:
     env:
-    - name: BUILDAH_FORMAT
-      value: oci
-    - name: STORAGE_DRIVER
-      value: vfs
-    - name: HERMETIC
-      value: $(params.HERMETIC)
-    - name: PREFETCH_INPUT
-      value: $(params.PREFETCH_INPUT)
-    - name: CONTEXT
-      value: $(params.CONTEXT)
-    - name: DOCKERFILE
-      value: $(params.DOCKERFILE)
-    - name: IMAGE
-      value: $(params.IMAGE)
-    - name: TLSVERIFY
-      value: $(params.TLSVERIFY)
-    - name: IMAGE_EXPIRES_AFTER
-      value: $(params.IMAGE_EXPIRES_AFTER)
+      - name: BUILDAH_FORMAT
+        value: oci
+      - name: STORAGE_DRIVER
+        value: vfs
+      - name: HERMETIC
+        value: $(params.HERMETIC)
+      - name: PREFETCH_INPUT
+        value: $(params.PREFETCH_INPUT)
+      - name: CONTEXT
+        value: $(params.CONTEXT)
+      - name: DOCKERFILE
+        value: $(params.DOCKERFILE)
+      - name: IMAGE
+        value: $(params.IMAGE)
+      - name: TLSVERIFY
+        value: $(params.TLSVERIFY)
+      - name: IMAGE_EXPIRES_AFTER
+        value: $(params.IMAGE_EXPIRES_AFTER)
   steps:
-  - image: $(params.BUILDER_IMAGE)
-    # per https://kubernetes.io/docs/concepts/containers/images/#imagepullpolicy-defaulting
-    # the cluster will set imagePullPolicy to IfNotPresent; our default param above specifies a digest
-    # also per direction from Ralph Bean, we want to use image digest based tags to use a cue to automation like dependabot or renovatebot to periodially submit pull requests that update the digest as new images are released.
-    name: build
-    computeResources:
-      limits:
-        memory: 4Gi
-      requests:
-        memory: 512Mi
-        cpu: 250m
-    env:
-    - name: COMMIT_SHA
-      value: $(params.COMMIT_SHA)
-    script: |
-      SOURCE_CODE_DIR=source
-      if [ -e "$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE" ]; then
-        dockerfile_path="$SOURCE_CODE_DIR/$CONTEXT/$DOCKERFILE"
-      elif [ -e "$SOURCE_CODE_DIR/$DOCKERFILE" ]; then
-        dockerfile_path="$SOURCE_CODE_DIR/$DOCKERFILE"
-      elif echo "$DOCKERFILE" | grep -q "^https\?://"; then
-        echo "Fetch Dockerfile from $DOCKERFILE"
-        dockerfile_path=$(mktemp --suffix=-Dockerfile)
-        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path" "$DOCKERFILE")
-        if [ $http_code != 200 ]; then
-          echo "No Dockerfile is fetched. Server responds $http_code"
+    - name: use-trusted-artifact
+      image: quay.io/redhat-appstudio/build-trusted-artifacts:latest@sha256:4581dc538a0bc1550466e8c7df3ee905e444eef3bc70a8ae765b568ab5a10588
+      args:
+        - use
+        - --store
+        - $(workspaces.source.path)
+        - $(params.ARTIFACTS[0])=$(workspaces.source.path)/source
```

Comment:
This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?

Rather than an array pram just use one parameter/result per artifact.

Reply:
Also this would remove the need to always run the prefetch step.

Reply:
> This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?
> 
> Rather than an array pram just use one parameter/result per artifact.

The number of artifacts produced by a task is unbounded, e.g. the git-clone task [create one artifact](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/git-clone/0.1/git-clone.yaml#L269), whereas prefetch-dependencies [produces two](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/prefetch-dependencies/0.1/prefetch-dependencies.yaml#L77-L78).

> Also this would remove the need to always run the prefetch step.

I'm curious, can you elaborate a bit?

Reply:
> > This kinda feels wrong to me. Wouldn't it make more sense to pass these in as explicit named parameters?
> > Rather than an array pram just use one parameter/result per artifact.
> 
> The number of artifacts produced by a task is unbounded, e.g. the git-clone task [create one artifact](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/git-clone/0.1/git-clone.yaml#L269), whereas prefetch-dependencies [produces two](https://github.com/zregvart/build-definitions/blob/945441d9219ac4233a022da9ece34aa8ef2601d3/task/prefetch-dependencies/0.1/prefetch-dependencies.yaml#L77-L78).

Each artifact is basically just a digest, so you would just name them, and then in the 'use' section specify the results you were reading from. E.g. git-clone would have a result called CHECKOUT_ARTIFACT, this would be pased in as a GIT_CHECKOUT param, and instead of `$(params.ARTIFACTS[0])=$(workspaces.source.path)/source` it would be `$(params.GIT_CHECKOUT=$(workspaces.source.path)/source`.

> 
> > Also this would remove the need to always run the prefetch step.
> 
> I'm curious, can you elaborate a bit?

I guess it would would still need to run, as I don't think Tekton allows subsequent tasks to run if a task is skipped, but you could abort early and write an empty file to the results, and have the artifact code ignore empty hashes.

Reply:
I think I can make `*_ARTIFACT` work, that will make the code easier to follow. I guess I was considering how we would tie the chain of artifacts easier in EC, having the same result/parameter name would help there.